File: crates/breez-sdk/bindings/langs/kotlin-multiplatform/breez-sdk-spark-kmp/src/commonTest/kotlin/breez_sdk_spark/webln/TypesTest.kt
package breez_sdk_spark.webln

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlin.test.assertFalse

class TypesTest {

    @Test
    fun testLnurlTypeValues() {
        // Verify all LnurlType enum values exist
        val pay = LnurlType.PAY
        val withdraw = LnurlType.WITHDRAW
        val auth = LnurlType.AUTH

        assertNotEquals(pay, withdraw)
        assertNotEquals(pay, auth)
        assertNotEquals(withdraw, auth)
    }

    @Test
    fun testLnurlRequestPayType() {
        val request = LnurlRequest(
            type = LnurlType.PAY,
            domain = "example.com",
            minAmountSats = 1000,
            maxAmountSats = 100000,
            metadata = """[["text/plain", "test"]]"""
        )

        assertEquals(LnurlType.PAY, request.type)
        assertEquals("example.com", request.domain)
        assertEquals(1000, request.minAmountSats)
        assertEquals(100000, request.maxAmountSats)
        assertEquals("""[["text/plain", "test"]]""", request.metadata)
        assertNull(request.defaultDescription)
    }

    @Test
    fun testLnurlRequestWithdrawType() {
        val request = LnurlRequest(
            type = LnurlType.WITHDRAW,
            domain = "service.com",
            minAmountSats = 100,
            maxAmountSats = 50000,
            defaultDescription = "Withdrawal"
        )

        assertEquals(LnurlType.WITHDRAW, request.type)
        assertEquals("service.com", request.domain)
        assertEquals(100, request.minAmountSats)
        assertEquals(50000, request.maxAmountSats)
        assertEquals("Withdrawal", request.defaultDescription)
        assertNull(request.metadata)
    }

    @Test
    fun testLnurlRequestAuthType() {
        val request = LnurlRequest(
            type = LnurlType.AUTH,
            domain = "auth.example.com"
        )

        assertEquals(LnurlType.AUTH, request.type)
        assertEquals("auth.example.com", request.domain)
        assertNull(request.minAmountSats)
        assertNull(request.maxAmountSats)
        assertNull(request.metadata)
        assertNull(request.defaultDescription)
    }

    @Test
    fun testLnurlUserResponseApproved() {
        val response = LnurlUserResponse(
            approved = true,
            amountSats = 5000,
            comment = "Thanks!"
        )

        assertTrue(response.approved)
        assertEquals(5000, response.amountSats)
        assertEquals("Thanks!", response.comment)
    }

    @Test
    fun testLnurlUserResponseRejected() {
        val response = LnurlUserResponse(approved = false)

        assertFalse(response.approved)
        assertNull(response.amountSats)
        assertNull(response.comment)
    }

    @Test
    fun testWebLnErrorCodes() {
        assertEquals("USER_REJECTED", WebLnErrorCode.USER_REJECTED)
        assertEquals("PROVIDER_NOT_ENABLED", WebLnErrorCode.PROVIDER_NOT_ENABLED)
        assertEquals("INSUFFICIENT_FUNDS", WebLnErrorCode.INSUFFICIENT_FUNDS)
        assertEquals("INVALID_PARAMS", WebLnErrorCode.INVALID_PARAMS)
        assertEquals("UNSUPPORTED_METHOD", WebLnErrorCode.UNSUPPORTED_METHOD)
        assertEquals("INTERNAL_ERROR", WebLnErrorCode.INTERNAL_ERROR)
    }
}

File: packages/react-native/src/webln/__tests__/types.test.ts
import { WebLnErrorCode, type LnurlRequest, type LnurlUserResponse } from '../types';

describe('WebLnErrorCode', () => {
  it('has all expected error codes', () => {
    expect(WebLnErrorCode.userRejected).toBe('USER_REJECTED');
    expect(WebLnErrorCode.providerNotEnabled).toBe('PROVIDER_NOT_ENABLED');
    expect(WebLnErrorCode.insufficientFunds).toBe('INSUFFICIENT_FUNDS');
    expect(WebLnErrorCode.invalidParams).toBe('INVALID_PARAMS');
    expect(WebLnErrorCode.unsupportedMethod).toBe('UNSUPPORTED_METHOD');
    expect(WebLnErrorCode.internalError).toBe('INTERNAL_ERROR');
  });
});

describe('LnurlRequest type', () => {
  it('can create pay request with all fields', () => {
    const request: LnurlRequest = {
      type: 'pay',
      domain: 'example.com',
      minAmountSats: 1000,
      maxAmountSats: 100000,
      metadata: '[["text/plain", "test"]]',
    };

    expect(request.type).toBe('pay');
    expect(request.domain).toBe('example.com');
    expect(request.minAmountSats).toBe(1000);
    expect(request.maxAmountSats).toBe(100000);
    expect(request.metadata).toBe('[["text/plain", "test"]]');
  });

  it('can create withdraw request with all fields', () => {
    const request: LnurlRequest = {
      type: 'withdraw',
      domain: 'service.com',
      minAmountSats: 100,
      maxAmountSats: 50000,
      defaultDescription: 'Withdrawal',
    };

    expect(request.type).toBe('withdraw');
    expect(request.domain).toBe('service.com');
    expect(request.minAmountSats).toBe(100);
    expect(request.maxAmountSats).toBe(50000);
    expect(request.defaultDescription).toBe('Withdrawal');
  });

  it('can create auth request with minimal fields', () => {
    const request: LnurlRequest = {
      type: 'auth',
      domain: 'auth.example.com',
    };

    expect(request.type).toBe('auth');
    expect(request.domain).toBe('auth.example.com');
    expect(request.minAmountSats).toBeUndefined();
  });
});

describe('LnurlUserResponse type', () => {
  it('can create approved response with amount', () => {
    const response: LnurlUserResponse = {
      approved: true,
      amountSats: 5000,
      comment: 'Thanks!',
    };

    expect(response.approved).toBe(true);
    expect(response.amountSats).toBe(5000);
    expect(response.comment).toBe('Thanks!');
  });

  it('can create rejected response', () => {
    const response: LnurlUserResponse = {
      approved: false,
    };

    expect(response.approved).toBe(false);
    expect(response.amountSats).toBeUndefined();
  });
});
